language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/semantictokens/SemanticTokensVisitor.java

```diff
@@ -0,0 +1,768 @@
+                                  int refModifiers) {
+        LinePosition startLine = node.lineRange().startLine();
+        SemanticToken semanticToken = new SemanticToken(startLine.line(), startLine.offset());
+        if (!semanticTokens.contains(semanticToken)) {
+            int length = node instanceof Token ? ((Token) node).text().trim().length() : node.textRange().length();
+            semanticToken.setProperties(length, type, modifiers);
+            semanticTokens.add(semanticToken);
+            if (processReferences) {
+                handleReferences(startLine, length, refType, refModifiers);
+            }
+        }
+    }
+
+    /**
+     * Handles references of the node that is located in the given position.
+     *
+     * @param linePosition Start position of the node
+     * @param length       Length to highlight
+     * @param type         Semantic token type's index
+     * @param modifiers    Semantic token type modifiers' index
+     */
+    private void handleReferences(LinePosition linePosition, int length, int type, int modifiers) {
+        Optional<SemanticModel> semanticModel;
+        try {
+            semanticModel = this.semanticTokensContext.currentSemanticModel();
+        } catch (Throwable e) {
+            return;
+        }
+        if (semanticModel.isEmpty()) {
+            return;
+        }
+
+        Optional<Document> docOptional = this.semanticTokensContext.currentDocument();
+        if (docOptional.isEmpty()) {
+            return;
+        }
+
+        Path path = this.semanticTokensContext.filePath().getFileName();
+        if (path == null) {
+            return;
+        }
+
+        semanticModel.get().references(docOptional.get(), linePosition).stream().filter(location ->
+                location.lineRange().filePath().equals(path.toString())).forEach(location -> {
+            LinePosition position = location.lineRange().startLine();
+            SemanticToken semanticToken = new SemanticToken(position.line(), position.offset());
+            if (!semanticTokens.contains(semanticToken)) {
+                semanticToken.setProperties(length, type, modifiers);
+                semanticTokens.add(semanticToken);
+            }
+        });
+    }
+
+    /**
+     * Represents semantic token data for a node.
+     */
+    static class SemanticToken implements Comparable<SemanticToken> {
+
+        private final int line;
+        private final int column;
+        private int length;
+        private int type;
+        private int modifiers;
+
+        private SemanticToken(int line, int column) {
+            this.line = line;
+            this.column = column;
+        }
+
+        private SemanticToken(int line, int column, int length, int type, int modifiers) {
+            this.line = line;
+            this.column = column;
+            this.length = length;
+            this.type = type;
+            this.modifiers = modifiers;
+        }
```

Comment:
This constructor seems to be unused. Do we have a requirement to keep it?

Reply:
Use this at https://github.com/ballerina-platform/ballerina-lang/pull/31841/files/25642f9a21c421ee73c0fc9870d26711e361bcfb#diff-d379b6637eb9a8380769712046ec25e219e3ca38bcdc3ccdc35e7d3b551ecb3bR130 and https://github.com/ballerina-platform/ballerina-lang/pull/31841/files/25642f9a21c421ee73c0fc9870d26711e361bcfb#diff-d379b6637eb9a8380769712046ec25e219e3ca38bcdc3ccdc35e7d3b551ecb3bR225